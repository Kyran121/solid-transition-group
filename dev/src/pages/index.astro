---
import Layout from "../layout.astro";
import Exmaple from "./kitchen-sink";
---

<Layout title="Kitchen Sink">
  <Exmaple client:idle />
</Layout>
