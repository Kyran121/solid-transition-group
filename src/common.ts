import { createMemo } from "solid-js";
import type { TransitionEvents } from "./Transition";

export function createClassnames(props: {
  name?: string;
  enterActiveClass?: string;
  enterClass?: string;
  enterToClass?: string;
  exitActiveClass?: string;
  exitClass?: string;
  exitToClass?: string;
  moveClass?: string;
}) {
  return createMemo(() => {
    const name = props.name || "s";
    return {
      enterActiveClasses: (props.enterActiveClass || name + "-enter-active").split(" "),
      enterClasses: (props.enterClass || name + "-enter").split(" "),
      enterToClasses: (props.enterToClass || name + "-enter-to").split(" "),
      exitActiveClasses: (props.exitActiveClass || name + "-exit-active").split(" "),
      exitClasses: (props.exitClass || name + "-exit").split(" "),
      exitToClasses: (props.exitToClass || name + "-exit-to").split(" "),
      moveClasses: (props.moveClass || name + "-move").split(" ")
    };
  });
}

// https://github.com/solidjs-community/solid-transition-group/issues/12
// for the css transition be triggered properly on firefox
// we need to wait for two frames before changeing classes
export function nextFrame(fn: () => void) {
  requestAnimationFrame(() => requestAnimationFrame(fn));
}

/**
 * Run an enter transition on an element - common for both Transition and TransitionGroup
 */
export function enterTransition(
  classnames: ReturnType<ReturnType<typeof createClassnames>>,
  events: TransitionEvents,
  el: Element,
  done?: VoidFunction
) {
  const { enterClasses, enterActiveClasses, enterToClasses } = classnames;

  // before the elements are added to the DOM
  events.onBeforeEnter && events.onBeforeEnter(el);

  el.classList.add(...enterClasses);
  el.classList.add(...enterActiveClasses);

  // after the microtask the elements will be added to the DOM
  // and onEnter will be called in the same frame
  queueMicrotask(() => {
    // Don't animate element if it's not in the DOM
    // This can happen when elements are changed under Suspense
    if (!el.parentNode) return done && done();

    events.onEnter && events.onEnter(el, () => endTransition());
  });

  nextFrame(() => {
    el.classList.remove(...enterClasses);
    el.classList.add(...enterToClasses);

    if (!events.onEnter || events.onEnter.length < 2) {
      el.addEventListener("transitionend", endTransition);
      el.addEventListener("animationend", endTransition);
    }
  });

  function endTransition(e?: Event) {
    if (!e || e.target === el) {
      el.removeEventListener("transitionend", endTransition);
      el.removeEventListener("animationend", endTransition);
      el.classList.remove(...enterActiveClasses);
      el.classList.remove(...enterToClasses);
      done && done(); // starts exit transition in "in-out" mode
      events.onAfterEnter && events.onAfterEnter(el);
    }
  }
}

/**
 * Run an exit transition on an element - common for both Transition and TransitionGroup
 */
export function exitTransition(
  classnames: ReturnType<ReturnType<typeof createClassnames>>,
  events: TransitionEvents,
  el: Element,
  done?: VoidFunction
) {
  const { exitClasses, exitActiveClasses, exitToClasses } = classnames;

  // Don't animate element if it's not in the DOM
  // This can happen when elements are changed under Suspense
  if (!el.parentNode) return done && done();

  events.onBeforeExit && events.onBeforeExit(el);

  el.classList.add(...exitClasses);
  el.classList.add(...exitActiveClasses);
  nextFrame(() => {
    el.classList.remove(...exitClasses);
    el.classList.add(...exitToClasses);
  });

  events.onExit && events.onExit(el, () => endTransition());

  if (!events.onExit || events.onExit.length < 2) {
    el.addEventListener("transitionend", endTransition);
    el.addEventListener("animationend", endTransition);
  }

  function endTransition(e?: Event) {
    if (!e || e.target === el) {
      el.removeEventListener("transitionend", endTransition);
      el.removeEventListener("animationend", endTransition);
      el.classList.remove(...exitActiveClasses);
      el.classList.remove(...exitToClasses);
      done && done(); // removes element from DOM, starts enter transition in "out-in" mode
      events.onAfterExit && events.onAfterExit(el);
    }
  }
}
