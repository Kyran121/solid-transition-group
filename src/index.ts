export { Transition } from "./Transition";
export type { TransitionProps, TransitionEvents } from "./Transition";

export { TransitionGroup } from "./TransitionGroup";
export type { TransitionGroupProps } from "./TransitionGroup";
