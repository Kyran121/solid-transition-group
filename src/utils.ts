import { createMemo } from "solid-js";

// https://github.com/solidjs-community/solid-transition-group/issues/12
// for the css transition be triggered properly on firefox
// we need to wait for two frames before changeing classes
export function nextFrame(fn: () => void) {
  requestAnimationFrame(() => requestAnimationFrame(fn));
}

export function createClassnames(props: {
  name?: string;
  enterActiveClass?: string;
  enterClass?: string;
  enterToClass?: string;
  exitActiveClass?: string;
  exitClass?: string;
  exitToClass?: string;
  moveClass?: string;
}) {
  return createMemo(() => {
    const name = props.name || "s";
    return {
      enterActiveClasses: (props.enterActiveClass || name + "-enter-active").split(" "),
      enterClasses: (props.enterClass || name + "-enter").split(" "),
      enterToClasses: (props.enterToClass || name + "-enter-to").split(" "),
      exitActiveClasses: (props.exitActiveClass || name + "-exit-active").split(" "),
      exitClasses: (props.exitClass || name + "-exit").split(" "),
      exitToClasses: (props.exitToClass || name + "-exit-to").split(" "),
      moveClasses: (props.moveClass || name + "-move").split(" ")
    };
  });
}
