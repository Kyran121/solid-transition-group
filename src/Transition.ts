import { type FlowComponent, type JSX } from "solid-js";
import { createClassnames, enterTransition, exitTransition } from "./common";
import { createSwitchTransition } from "@solid-primitives/transition-group";
import { resolveFirst } from "@solid-primitives/refs";

export type TransitionEvents = {
  /**
   * Function called before the enter transition starts — before enter classes are applied.
   */
  onBeforeEnter?: (el: Element) => void;
  /**
   * Function called when the enter transition starts — after `enterToClass` class is applied.
   * Call `done` to end the transition - removes enter classes and calls `onAfterEnter`.
   */
  onEnter?: (el: Element, done: () => void) => void;
  /**
   * Function called after the enter transition ends — after all enter classes are removed.
   */
  onAfterEnter?: (el: Element) => void;
  /**
   * Function called before the exit transition starts — before exit classes are applied.
   */
  onBeforeExit?: (el: Element) => void;
  /**
   * Function called when the exit transition starts — after `exitToClass` class is applied.
   * Call `done` to end the transition - removes exit classes, calls `onAfterExit` and removes the element from the DOM.
   */
  onExit?: (el: Element, done: () => void) => void;
  /**
   * Function called after the exit transition ends — after all exit classes are removed.
   */
  onAfterExit?: (el: Element) => void;
};

export type TransitionProps = TransitionEvents & {
  /**
   * Used to automatically generate transition CSS class names.
   * e.g. `name: 'fade'` will auto expand to `.fade-enter`, `.fade-enter-active`, etc.
   * Defaults to `"s"`.
   */
  name?: string;
  /**
   * CSS class applied to the entering element for the entire duration of the enter transition.
   */
  enterActiveClass?: string;
  /**
   * CSS class applied to the entering element at the start of the enter transition, and removed the frame after.
   */
  enterClass?: string;
  /**
   * CSS class applied to the entering element after the enter transition starts.
   */
  enterToClass?: string;
  /**
   * CSS class applied to the exiting element for the entire duration of the exit transition.
   */
  exitActiveClass?: string;
  /**
   * CSS class applied to the exiting element at the start of the exit transition, and removed the frame after.
   */
  exitClass?: string;
  /**
   * CSS class applied to the exiting element after the exit transition starts.
   */
  exitToClass?: string;
  /**
   * Whether to apply transition on initial render. Defaults to `false`.
   */
  appear?: boolean;
  /**
   * Controls the timing sequence of leaving/entering transitions. Available modes are `"outin"` and `"inout"`;
   * Defaults to simultaneous.
   */
  mode?: "inout" | "outin";
};

const TRANSITION_MODE_MAP = {
  inout: "in-out",
  outin: "out-in"
} as const;

export const Transition: FlowComponent<TransitionProps> = props => {
  const classnames = createClassnames(props);

  return createSwitchTransition(
    resolveFirst(() => props.children),
    {
      mode: TRANSITION_MODE_MAP[props.mode!],
      appear: props.appear,
      onEnter: enterTransition.bind(null, classnames, props),
      onExit: exitTransition.bind(null, classnames, props)
    }
  ) as unknown as JSX.Element;
};
