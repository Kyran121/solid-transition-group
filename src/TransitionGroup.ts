import { type JSX, type FlowComponent } from "solid-js";
import { createClassnames, enterTransition, exitTransition } from "./common";
import { createListTransition } from "@solid-primitives/transition-group";
import { resolveElements } from "@solid-primitives/refs";
import type { TransitionProps } from "./Transition";

/**
 * Props for the {@link TransitionGroup} component.
 */
export type TransitionGroupProps = Omit<TransitionProps, "mode"> & {
  /**
   * CSS class applied to the moving elements for the entire duration of the move transition.
   * Defaults to `"s-move"`.
   */
  moveClass?: string;
};

/**
 * The `<TransitionGroup>` component lets you apply enter and leave animations on elements passed to `props.children`.
 *
 * It supports transitioning multiple elements at a time and moving elements around.
 *
 * @param props {@link TransitionGroupProps}
 */
export const TransitionGroup: FlowComponent<TransitionGroupProps> = props => {
  const classnames = createClassnames(props);

  return createListTransition(resolveElements(() => props.children).toArray, {
    appear: props.appear,
    exitMethod: "keep-index",
    onChange({ added, removed, finishRemoved, unchanged, list }) {
      const classes = classnames();

      // ENTER
      for (const el of added) {
        enterTransition(classes, props, el);
      }

      // MOVE
      const elsToMove: { el: HTMLElement | SVGElement; rect: DOMRect }[] = [];
      // get rects of elements before the changes to the DOM
      for (const el of list) {
        if (el.isConnected && (el instanceof HTMLElement || el instanceof SVGElement)) {
          elsToMove.push({ el, rect: el.getBoundingClientRect() });
        }
      }

      // wait for changes to be rendered
      queueMicrotask(() => {
        const moved: (HTMLElement | SVGElement)[] = [];

        for (const { el, rect } of elsToMove) {
          if (el.isConnected) {
            const newRect = el.getBoundingClientRect(),
              dX = rect.left - newRect.left,
              dY = rect.top - newRect.top;
            if (dX || dY) {
              // set els to their old position before transition
              el.style.transform = `translate(${dX}px, ${dY}px)`;
              el.style.transitionDuration = "0s";
              moved.push(el);
            }
          }
        }

        document.body.offsetHeight; // force reflow

        for (const el of moved) {
          el.classList.add(...classes.moveClasses);

          // clear transition - els will move to their new position
          el.style.transform = el.style.transitionDuration = "";

          function endTransition(e: Event) {
            if (e.target === el || /transform$/.test((e as TransitionEvent).propertyName)) {
              el.removeEventListener("transitionend", endTransition);
              el.classList.remove(...classes.moveClasses);
            }
          }
          el.addEventListener("transitionend", endTransition);
        }
      });

      // EXIT
      for (const el of removed) {
        exitTransition(classes, props, el, () => finishRemoved([el]));
      }
    }
  }) as unknown as JSX.Element;
};
