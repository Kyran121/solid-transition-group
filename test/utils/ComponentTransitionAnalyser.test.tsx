import { describe, it } from "vitest";
import { createSignal, Show, type Component } from "solid-js";
import { Transition } from "../../src";
import ComponentTransitionAnalyser, {
  type TransitionComponent
} from "./ComponentTransitionAnalyser";
import { fireEvent } from "@solidjs/testing-library";
import "./ComponentTransitionAnalyser.css";

type TransitionProps = {
  show?: boolean;
  enterDuration?: number;
  exitDuration?: number;
};

describe.concurrent("ComponentTransitionAnalyser", () => {
  describe("analyseTransitionActivity", () => {
    it("throws error when transitioning element is missing duration (entering transition)", async ({
      expect
    }) => {
      const TransitionComponent = createTransitionComponent({
        show: false,
        enterDuration: 9999
      });

      const componentTransitionAnalyser = new ComponentTransitionAnalyser(TransitionComponent);
      componentTransitionAnalyser.addTransitionTrigger(clickToggleButtonWaitingDuration(75));

      expect(componentTransitionAnalyser.analyseTransitionActivity()).rejects.toThrow(
        "Unable to extract transitionDuration style from transition container child!"
      );
    });

    it("throws error when transition fails to take specified amount of time", async ({
      expect
    }) => {
      const enterDuration = 150;

      const TransitionComponent = createTransitionComponent({
        show: false,
        enterDuration: 75
      });

      const componentTransitionAnalyser = new ComponentTransitionAnalyser(TransitionComponent);
      componentTransitionAnalyser.addTransitionTrigger(
        clickToggleButtonWaitingDuration(enterDuration)
      );

      expect(componentTransitionAnalyser.analyseTransitionActivity()).rejects.toThrow(
        `Transition #1 failed to take at least ${enterDuration}ms`
      );
    });

    it("correctly analyses activity for an element which enters", async ({ expect }) => {
      const enterDuration = 75;

      const TransitionComponent = createTransitionComponent({
        show: false,
        enterDuration
      });

      const componentTransitionAnalyser = new ComponentTransitionAnalyser(TransitionComponent);
      componentTransitionAnalyser.addTransitionTrigger(
        clickToggleButtonWaitingDuration(enterDuration)
      );

      const activityReport = await componentTransitionAnalyser.analyseTransitionActivity();
      expect(activityReport).toMatchSnapshot();
    });

    it("correctly analyses activity for an element which leaves", async ({ expect }) => {
      const exitDuration = 75;

      const TransitionComponent = createTransitionComponent({
        show: true,
        exitDuration
      });

      const componentTransitionAnalyser = new ComponentTransitionAnalyser(TransitionComponent);
      componentTransitionAnalyser.addTransitionTrigger(
        clickToggleButtonWaitingDuration(exitDuration)
      );

      const activityReport = await componentTransitionAnalyser.analyseTransitionActivity();
      expect(activityReport).toMatchSnapshot();
    });

    it("correctly analyses activity for element which enters and leaves", async ({ expect }) => {
      const enterDuration = 75;
      const exitDuration = 100;

      const TransitionComponent = createTransitionComponent({
        show: false,
        enterDuration,
        exitDuration
      });

      const componentTransitionAnalyser = new ComponentTransitionAnalyser(TransitionComponent);
      componentTransitionAnalyser.addTransitionTrigger(
        clickToggleButtonWaitingDuration(enterDuration)
      );
      componentTransitionAnalyser.addTransitionTrigger(
        clickToggleButtonWaitingDuration(exitDuration)
      );

      const activityReport = await componentTransitionAnalyser.analyseTransitionActivity();
      expect(activityReport).toMatchSnapshot();
    });

    function createTransitionComponent(props: TransitionProps): Component {
      return () => {
        const [show, setShow] = createSignal(props.show);

        return (
          <>
            <div data-testid="transition-container">
              <Transition
                enterActiveClass={`duration-${props.enterDuration} enter-active`}
                enterClass="opacity-0 enter"
                enterToClass="opacity-100 enter-to"
                exitActiveClass={`duration-${props.exitDuration} exit-active`}
                exitClass="opacity-100 exit"
                exitToClass="opacity-0 exit-to"
              >
                <Show when={show()}>
                  <div>
                    <span>Hello!</span>
                  </div>
                </Show>
              </Transition>
            </div>
            <button data-testid="toggle" onClick={() => setShow(!show())}>
              Toggle
            </button>
          </>
        );
      };
    }

    function clickToggleButtonWaitingDuration(expectedDuration: number) {
      const execute = (tools: TransitionComponent) => fireEvent.click(tools.getByTestId("toggle"));
      return { execute, expectedDuration };
    }
  });
});
