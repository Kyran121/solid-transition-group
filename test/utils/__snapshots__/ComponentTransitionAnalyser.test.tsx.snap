// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ComponentTransitionAnalyser > analyseTransitionActivity > correctly analyses activity for an element which enters 1`] = `
"Render 1:
    +  <div
    +    class="duration-75 enter-active opacity-0 enter"
    +  >
    +    <span>
    +      Hello!
    +    </span>
    +  </div>

Render 2:
    -  class="duration-75 enter-active opacity-0 enter"
    +  class="duration-75 enter-active opacity-100 enter-to"

Render 3: Transition took at least 75ms
    -  class="duration-75 enter-active opacity-100 enter-to"
    +  class"
`;

exports[`ComponentTransitionAnalyser > analyseTransitionActivity > correctly analyses activity for an element which leaves 1`] = `
"Render 1:
    -  <div>
    +  <div
    +    class="duration-75 exit-active opacity-100 exit"
    +  >

Render 2:
    -  class="duration-75 exit-active opacity-100 exit"
    +  class="duration-75 exit-active opacity-0 exit-to"

Render 3: Transition took at least 75ms
    -  <div
    -    class="duration-75 exit-active opacity-0 exit-to"
    -  >
    -    <span>
    -      Hello!
    -    </span>
    -  </div>"
`;

exports[`ComponentTransitionAnalyser > analyseTransitionActivity > correctly analyses activity for element which enters and leaves 1`] = `
"Render 1:
    +  <div
    +    class="duration-75 enter-active opacity-0 enter"
    +  >
    +    <span>
    +      Hello!
    +    </span>
    +  </div>

Render 2:
    -  class="duration-75 enter-active opacity-0 enter"
    +  class="duration-75 enter-active opacity-100 enter-to"

Render 3: Transition took at least 75ms
    -  class="duration-75 enter-active opacity-100 enter-to"
    +  class

Render 4:
    -  class
    +  class="duration-100 exit-active opacity-100 exit"

Render 5:
    -  class="duration-100 exit-active opacity-100 exit"
    +  class="duration-100 exit-active opacity-0 exit-to"

Render 6: Transition took at least 100ms
    -  <div
    -    class="duration-100 exit-active opacity-0 exit-to"
    -  >
    -    <span>
    -      Hello!
    -    </span>
    -  </div>"
`;
